### PR TITLE
[merged] Introduce registry inspect methods

### PR DESCRIPTION
**Pull request by baude**
Add two classes and several new methods for inspecting
registries and images.  These methods are helpful
for:

* taking user input (i.e. image names, partial image names) and determining
the fully qualified names.
* determining the fq name when the registry is omitted by the user.
* obtaining image manifests
* mimicing skopeo inspect

### CODE REVIEW

File: Atomic/atomic.py [hunk 1]
```diff
@@ -700,6 +702,18 @@ def load_local_tokens():
             pass
         return tokens
 
+    def get_fq_image_name(self, input_image):
+        registry, repo, image, tag = util.decompose(input_image)
+        if not image:
+            raise ValueError('Error parsing input: "{}" invalid'.format(input_image))
+        if all([True if x else False for x in [registry, repo, image, tag]]):
+            return input_image
+        if all([True if x else False for x in [registry, image, tag]]):
```
Comment (rhatdan) [inline Atomic/atomic.py]: Why do the first one?  Doing just this block should be enough?

**Comment by rhatdan** [comment]
@runcom PTAL. Since he is doing some of the stuff skopeo does for remote viewing...

**Comment by rhatdan** [comment]
Are you handling the case where there are multiple registires?

atomic pull rhel7tools

Does this work there?

File: Atomic/atomic.py (hunk 1 above)
Comment (baude) [inline Atomic/atomic.py]: Yup! Fixed.

File: tests/unit/test_discovery.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
+import unittest
+from Atomic import util
+from Atomic import discovery
+
+
+class TestAtomicUtil(unittest.TestCase):
+    IMAGE = 'docker.io/library/busybox:latest'
```
Comment (runcom) [inline tests/unit/test_discovery.py]: is this working with a digest? e.g. docker.io/library/busybox@sha256:%HEX%

**Comment by runcom** [comment]
patch LGTM

> Are you handling the case where there are multiple registires?
> 
> atomic pull rhel7tools
> 
> Does this work there?

The loop over the available registries in docker should be here https://github.com/projectatomic/atomic/pull/687/files#diff-1e7bfb73126e6f7e89b71bbe478132d3R249 right Brent?

**Comment by baude** [comment]
yeah that is the loop.  I was trying to understand if that is what @rhatdan wanted or something different.  thanks for the review @runcom

**Comment by rhatdan** [comment]
Right as long as rhel7tools ends up being something like 
redhat-registry.redhat.com/rhel7tools:latest

I am fine with this patch.

**Comment by baude** [comment]
Current output for this looks like:

```
[bbaude@localhost atomic (inspect)]$ sudo ./atomic pull rhel7
Image rhel7 is being pulled to docker ...
Trying docker.io/library/rhel7:latest
Trying registry.access.redhat.com/rhel7:latest
Pulling registry.access.redhat.com/rhel7:latest ...
Getting image source manifest
Getting image source configuration
Uploading blob sha256:30cf2e26a24f2a8426cbe8444f8af2ecb7023bd468b05c1b6fd0b2797b0f9ff9
 5.97 MB / ? [-----------------------------
```

File: Atomic/discovery.py [hunk 3]
```diff
@@ -0,0 +1,406 @@
...
+import hashlib
+
+
+try:
+    from urllib.request import parse_http_list, parse_keqv_list
+except ImportError:
+    from urllib2 import parse_http_list, parse_keqv_list #pylint: disable=import-error
+
+
+def jose_base64_urldecode(in_str):
```
Comment (cgwalters) [inline Atomic/discovery.py]: Who's "jose"?  :smile:   Did this code come from somewhere else?

**Comment by baude** [comment]
At the risk of merging my own code ... but doing so to unblock CAHC

@rh-atomic-bot r+

File: Atomic/discovery.py (hunk 3 above)
Comment (baude) [inline Atomic/discovery.py]: @cgwalters it mimics some golang code

**Pull request closed without merging**